File: src/com/chbachman/cminus/representation/InitBlock.java
package com.chbachman.cminus.representation;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.function.*;
import com.chbachman.cminus.representation.statement.Assignment;
import com.chbachman.cminus.representation.statement.Return;
import com.chbachman.cminus.representation.value.Variable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Chandler on 5/5/17.
 */
public class InitBlock extends Function {

    public InitBlock(CMinusParser.InitBlockContext ctx, Struct parent, Scope scope) {
        super(parent.type(), "init" + parent.name, new ArrayList<>());
        parameters = ParameterList.parse(ctx.parameterList());

        scope.pushScope(this);

        // Create the temp variable to return.
        Variable toInit = new Variable("created" + parent.name, parent.type());
        scope.setThis(toInit, parent);
        this.statements.add(toInit);

        // Add the declared variables if declared inline.
        for(Variable v: parent.variables) {
            if (v.value.isPresent()) {
                this.statements.add(new Assignment(toInit.name + "." + v.name, v.value.get()));
            }
        }

        // Create the rest of the codeblock.
        this.statements.addAll(CodeBlock.parse(ctx.codeBlock(), scope));

        // Add the return value.
        this.statements.add(new Return(toInit));

        scope.popScope();
    }

    @Override
    public void setupScope(Scope scope) {
        for(Variable v : parameters) {
            scope.addVariable(v);
        }
    }
}

File: src/main/java/com/chbachman/cminus/representation/InitBlock.kt
package com.chbachman.cminus.representation

import com.chbachman.cminus.gen.CMinusParser
import com.chbachman.cminus.representation.function.CodeBlock
import com.chbachman.cminus.representation.function.Function
import com.chbachman.cminus.representation.function.ParameterList
import com.chbachman.cminus.representation.statement.Assignment
import com.chbachman.cminus.representation.statement.Return
import com.chbachman.cminus.representation.value.Variable

/**
 * Created by Chandler on 5/5/17.
 * Representation of Init Block.
 */
class InitBlock(ctx: CMinusParser.InitBlockContext, parent: Struct, scope: Scope) : Function(parent.type, "init" + parent.name, ParameterList.parse(ctx.parameterList())) {

    init {

        scope.pushScope(this)

        // Create the temp variable to return.
        val toInit = Variable("created" + parent.name, parent.type)
        scope.setThis(toInit, parent)
        this.statements.add(toInit)

        // Add the declared variables if declared inline.
        for (v in parent.getVariables()) {
            this.statements.add(Assignment(toInit.name + "." + v.name, v.value))
        }

        // Create the rest of the codeblock.
        this.statements.addAll(CodeBlock.Companion.parse(ctx.codeBlock(), scope))

        // Add the return value.
        this.statements.add(Return(toInit))

        scope.popScope()
    }

    override fun setupScope(scope: Scope) {
        for (v in this.parameters) {
            scope.addVariable(v)
        }
    }
}
